plugins/tc/tbf/main.go

```diff
@@ -0,0 +1,235 @@
+		}
+		res, err := version.NewResult(conf.CNIVersion, resultBytes)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse prevResult: %v", err)
+		}
+		conf.RawPrevResult = nil
+		conf.PrevResult, err = current.NewResultFromResult(res)
+		if err != nil {
+			return nil, fmt.Errorf("could not convert result to current version: %v", err)
+		}
+	}
+	// End previous result parsing
+
+	switch {
+	case conf.Burst < 0 || conf.Rate < 0:
+		return nil, fmt.Errorf("rate and burst must be a positive integer")
+	case conf.Burst == 0 && conf.Rate != 0:
+		return nil, fmt.Errorf("if rate is set, burst must also be set")
+	case conf.Rate == 0 && conf.Burst != 0:
+		return nil, fmt.Errorf("if burst is set, rate must also be set")
+	}
+
+	return &conf, nil
+
+}
+
+func getRoutableHostIF(containerIP net.IP) string {
+	routes, err := netlink.RouteGet(containerIP)
+	if err != nil {
+		return ""
+	}
+
+	for _, route := range routes {
+		link, err := netlink.LinkByIndex(route.LinkIndex)
+		if err != nil || link.Attrs().Name == "lo" {
+			continue
+		}
+
+		return link.Attrs().Name
+	}
+
+	return ""
+}
+
+// cmdAdd is called for ADD requests
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	if conf.Rate == 0 && conf.Burst == 0 {
+		return types.PrintResult(conf.PrevResult, conf.CNIVersion)
+	}
+
+	if conf.PrevResult == nil {
+		return fmt.Errorf("must be called as chained plugin")
+	}
+
+	containerIPConfigs := make([]*current.IPConfig, 0, len(conf.PrevResult.IPs))
+
+	for _, ip := range conf.PrevResult.IPs {
+		if ip.Interface == nil {
+			continue
+		}
+		intIdx := *ip.Interface
+		// Every IP is indexed in to the interfaces array, with "-1" standing
+		// for an unknown interface (which we'll assume to be Container-side
+		// Skip all IPs we know belong to an interface with the wrong name.
+		if intIdx >= 0 && intIdx < len(conf.PrevResult.Interfaces) && conf.PrevResult.Interfaces[intIdx].Name != args.IfName {
+			continue
+		}
+		containerIPConfigs = append(containerIPConfigs, ip)
+	}
+
+	if len(containerIPConfigs) == 0 {
+		return fmt.Errorf("got no container IPs")
+	}
+
+	for _, cip := range containerIPConfigs {
+		containerNs, err := ns.GetNS(conf.PrevResult.Interfaces[*cip.Interface].Sandbox)
+		if err != nil {
+			return err
+		}
+		err = containerNs.Do(func(_ ns.NetNS) error {
+			return CreateIngressQdisc(conf.Rate, conf.Burst, conf.PrevResult.Interfaces[*cip.Interface].Name)
+		})
+		if err != nil {
+			return err
+		}
+
+		for pidx, prevIface := range conf.PrevResult.Interfaces {
+			if *cip.Interface == pidx {
+				continue
+			}
+			nonContainerNetNs, err := ns.GetNS(prevIface.Sandbox)
+			if err != nil {
+				return err
+			}
+			err = nonContainerNetNs.Do(func(netNS ns.NetNS) error {
+				nonContainerIfaceName := getRoutableHostIF(cip.Address.IP)
+				if nonContainerIfaceName != "" {
+					ifbDeviceName := os.Getenv("CNI_IFNAME")
```

Comment:
Ah, also, this is not correct.  According to [the CNI spec](https://github.com/containernetworking/cni/blob/869505bebdf96fc1176bc29ad0893b867e717f6a/SPEC.md), `CNI_IFNAME` is the desired name for the primary network interface *inside the container*, i.e. `eth0`.  It is guaranteed unique only inside the container network namespace, not the host.  You'll need to generate a unique name for the host-side interface -- perhaps by hashing the `(network name, container id)` pair (see [here for some considerations](https://github.com/containernetworking/cni/blob/869505bebdf96fc1176bc29ad0893b867e717f6a/SPEC.md#general-considerations)), using a random name (see [here](https://github.com/containernetworking/plugins/blob/92c634042c3888d942199c396c1d6f42d82785f3/pkg/ip/link_linux.go#L87-L96)), or by using an auto-incrementing integer on the system.  You may need to store this name on the local filesystem in order to properly clean it up when `DEL` is called.

Reply:
I like the idea of generating a deterministic unique name for the host side interface. I have pushed a change to compute the sha1 of the (network name, container id) and only using the first 15 characters of the hex digest.
Thus on cmdDel i am able to determine the device name to delete.

On the off chance that the sha1 conflicts with an interface that already exists, this plugin will fail... If this is an issue then I can change to the random name approach + stored state. (i.e. if the random name conflicts, generate a new random name until no conflict)